apiserver/cloud/instance_information.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cloud
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/stateenvirons"
+	names "gopkg.in/juju/names.v2"
+)
+
+// EnvironConfigGetter implements environs.EnvironConfigGetter
+// in terms of a *state.State.
+type cloudEnvironConfigGetter struct {
+	*state.State
+	region string
+}
+
+// CloudSpec implements environs.EnvironConfigGetter.
+func (g cloudEnvironConfigGetter) CloudSpec(tag names.ModelTag) (environs.CloudSpec, error) {
+	model, err := g.GetModel(tag)
+	if err != nil {
+		return environs.CloudSpec{}, errors.Trace(err)
+	}
+	cloudName := model.Cloud()
+	regionName := g.region
+	credentialTag, _ := model.CloudCredential()
+	return stateenvirons.CloudSpec(g.State, cloudName, regionName, credentialTag)
+}
+
+// InstanceTypes returns instance type information for the cloud and region
+// in which the current model is deployed.
+func (api *CloudAPI) InstanceTypes(cons params.CloudInstanceTypesConstraints) (params.InstanceTypesResults, error) {
+	c, err := api.backend.ControllerModel()
+	if err != nil {
+		return params.InstanceTypesResults{}, errors.Trace(err)
+	}
+	modelTag := c.ModelTag()
+	result := make([]params.InstanceTypesResult, len(cons.Constraints))
+	stShim := api.backend.(*stateShim)
+	st := stShim.State
+	// TODO(perrito666) Cache the results to avoid excessive querying of the cloud.
+	for i, cons := range cons.Constraints {
+		value := constraints.Value{}
+		if cons.Constraints != nil {
+			value = *cons.Constraints
+		}
+		backend := cloudEnvironConfigGetter{
```

Comment:
Please parse the cons.CloudTag here, and check that it is the same as the controller model's cloud. We may relax this later, but it's better to be strict to avoid having to deal with sloppily written clients later.